### ISSUE TITLE
Studio: failed to save commands

### ISSUE THREAD

**Issue by lukeapage** [cypress-io/cypress#14767]
### Current behavior
![2021-01-27 08 50 54](https://user-images.githubusercontent.com/309321/105960401-71b81500-607d-11eb-8860-48053b97340b.gif)

### Test code to reproduce

I created a new repo, added cypress, did cypress open, deleted examples, created a new spec:
```
describe('soak test', () => {
 	it('logs in', () => { 
		cy.visit('internal url'); 
	}); 
});
```

Then I opened studio, typed into an input and clicked save.

The page refreshes so I needed to use animation capture to see the error.

How can I provide more details on the exact error? I don't see anything in the debug log but I will attach it here.

### Versions

6.3.0
windows 10

**Comment by lukeapage** [comment]
[debug.txt](https://github.com/cypress-io/cypress/files/5878757/debug.txt)

**Comment by jennifer-shehane** [comment]
Is there anything printed to the DevTools console when this error occurs?

**Comment by lukeapage** [comment]
No, I checked that

**Comment by jennifer-shehane** [comment]
Ok, we'll need some more information on how to reproduce this because we're not sure what would cause this error exactly - to print that it could not save the commands, but then to also refresh the commands (which only happens when a save of the commands is successful).

**Comment by elio-maggini** [comment]
I am seeing this on Mac as well. All I am trying to do is use Studio to visit the login page, nothing more.

STR:
1) Launch ng e2e on port 8100 (Version 88.0.4324.96 (Official Build) (x86_64))
2) Hover on "add new test" next to test name
3) click add new test
4) click get started (which shows every time)
5) enter a valid url (http://localhost:8100/login
6) click save
7) Get error message

macOS Catalina 10.15.7
Processor: 2.6 GHz 6-core Intel Core i7
Memory: 16GB 2667 MHz DDR4
Graphics AMD Radeon Pro 5300M 4GB
Intel UHD Graphics 630 1536 MB
-----
System report
Model Name:	MacBook Pro
  Model Identifier:	MacBookPro16,1
  Processor Name:	6-Core Intel Core i7
  Processor Speed:	2.6 GHz
  Number of Processors:	1
  Total Number of Cores:	6
  L2 Cache (per Core):	256 KB
  L3 Cache:	12 MB
  Hyper-Threading Technology:	Enabled
  Memory:	16 GB
  Boot ROM Version:	1554.60.15.0.0 (iBridge: 18.16.13030.0.0,0)
  Serial Number (system):	XXX
  Hardware UUID:	XXX
  Activation Lock Status:	Disabled

New Testpassed
STUDIO COMMANDS
1
visit
http://localhost:8100/login
(xhr)
GET 200 /sockjs-node/info?t=1611872393151
CypressError
Cypress was unable to save these commands to your spec file. Cypress Studio is still in beta and the team is working hard to resolve issues like this. To help us fix this issue more quickly, you can provide us with more information by clicking 'Learn more' below.Learn more

**Comment by jennifer-shehane** [comment]
Do either of you use test configuration in the test or suite you're trying to add commands to? If so, I've identified an issue with saving commands in that case. 

Something like below?

```js
// suite config for retries
describe('tests', { retries: 2 }, () => {
  it('test 1', () => {
    cy.visit('index.html')
  })
})

describe('tests', () => {
  // test config for repsonseTimeout
  it('test 1', { repsonseTimeout: 2 }, () => {
    cy.visit('index.html')
  })
})
```

**Comment by lukeapage** [comment]
My test case file was exactly as I put in the initial description and it was a brand new setup - all I did was enable studio.

**Comment by elio-maggini** [comment]
@jennifer-shehane Hopefully the other ticket will also fix this one. A little more info on this version after reading other ticket:

1) Save command fails no mater if adding  new test to suite or new command to test.
2) It is worth drawing attention to my install info above which indicates protractor has been removed and the cypress "stuff" resides in the e2e directory rather than the "cypress" directory from a standard install.

"@nrwl/cypress": "11.2.0",
"@nrwl/workspace": "11.2.0",
 "cypress": "6.3.0",
 
 If helpful, I could also try doing a full install on dev PC to see if the result is any different on Windows.

**Comment by panzarino** [comment]
@lukeapage Are you using any special characters or spaces in the path to your test files (#14659)? If so, that's been fixed and released with `6.4.0`. We've also fixed a number of other issues in `6.4.0`, could you test with that version and let us know if the problem is still happening?

**Comment by diggabyte** [comment]
Also seeing this error. 

Relatedly, it would be a nice feature to just display the recorded commands in a modal so we could just copy/paste it and avoid mutating the filesystem (might be particularly useful for those with custom suite functions or non-standard setups).

**Setup:**
* Cypress `6.4.0` (Mac)
* Electron `87.0.4280.141`
* No errors in console
* No info in the debug log
* No test configuration in suite

**Relevant packages**:
```
"@cypress/webpack-preprocessor": "^5.5.0",
"cypress": "6.4.0",
"cypress-file-upload": "^4.1.1",
"cypress-localstorage-commands": "^1.1.6",
"cypress-plugin-snapshots": "^1.4.4",
"ts-loader": "^5",
"typescript": "3.7.3",
```

**Webpack preprocessor options:**
```
extensions: ['.ts', '.tsx', '.js'],
test: /\.tsx?$/,
loader: 'ts-loader',
options: { transpileOnly: true },
```

**Relevant Config:**
```
{
  "baseUrl": "http://localhost:9000",
  "chromeWebSecurity": false,
  "video": false,
  "videoUploadOnPasses": false,
  "screenshotOnRunFailure": false,
  "supportFile": "cypress/support/index.ts",
  "defaultCommandTimeout": 10000,
  "requestTimeout":  30000,
  "modifyObstructiveCode": false,
  "experimentalSourceRewriting": false,
  "experimentalStudio": true,
  "numTestsKeptInMemory": 50,
  "ignoreTestFiles": [
    "**/__snapshots__/*",
    "**/__image_snapshots__/*"
  ],
  "trashAssetsBeforeRuns": true,
  "viewportHeight": 900,
  "viewportWidth": 1200,
  "env": {
    "cypress-plugin-snapshots": {
      "autoCleanUp": true,
      "separator": "_"
    },
    "failSilently": false
  }
}
```

**Comment by diggabyte** [comment]
I was also seeing this error, but we use a custom `suite` function that does some setup before work (reset db, authenticate as a specific user type). When I removed the custom suite method and used a standard `describe` it was able to write the commands.

Given that there may be non-standard setups like this, it would be a useful feature to simply display the commands in a modal so the user could copy/paste them rather than trying to parse the file and write to disk.

**Setup:**
Cypress 6.4.0 (Mac)
Electron 87.0.4280.141

**Before (using our custom suite function):**
```
suite('ADMIN', 'Suite Name', () => {
  it('does', () => {
    cy.visit('/#/home')
  });
});
```

**After (standard describe):**
```
describe('suite', () => {
  it('logs in', () => {
    cy.visit('/#/login')
    /* ==== Generated with Cypress Studio ==== */
    cy.get('.emailInput').type('asdasd');
    cy.get('.passwordInput').type('asdasd');
    /* ==== End Cypress Studio ==== */
  });
});
```

**Custom suite function (basically extends `describe`):**
```
const suite = (
  user: UserType,
  suiteName: string,
  fn: RunnerFn,
): Suite => {
  return describe(`Suite: ${suiteName}`, function suiteDescribe() {
    before(() => {
      resetDatabase();
      logInAs(user);
    });
    beforeEach(() => {
      cy.restoreLocalStorage();
    });
    fn.apply(this);
  });
};
```

**Relevant Packages:**
```
"@cypress/webpack-preprocessor": "^5.5.0",
"cypress": "6.4.0",
"cypress-file-upload": "^4.1.1",
"cypress-localstorage-commands": "^1.1.6",
"cypress-plugin-snapshots": "^1.4.4",
"ts-loader": "^5",
"typescript": "3.7.3",
```

**Webpack preprocessor options:**
```
extensions: ['.ts', '.tsx', '.js'],
test: /\.tsx?$/,
loader: 'ts-loader',
options: { transpileOnly: true },
```

**Comment by elio-maggini** [comment]
@diggabyte and @panzarino  ... the ability to copy and paste is the most brilliant work around. I am current using the manual copy method which is a real PITA when running `ng e2e --watch`

**Comment by diggabyte** [comment]
Worth noting, in my [custom suite comment](https://github.com/cypress-io/cypress/issues/14767#issuecomment-776912605) above, I've discovered I can record / save to an existing test, but adding a new suite is what fails (probably due to the non-standard suite method). Still.. ability to display the commands (or copy them to clipboard by clicking a button) would be a great feature!

**Comment by jennifer-shehane** [comment]
@diggabyte I'm not able to recreate this failure to save commands with the custom suite definition. Is this issue still happening for you in the latest version?

<img width="715" alt="Screen Shot 2021-02-27 at 9 40 37 AM" src="https://user-images.githubusercontent.com/1271364/109390476-e03bfd00-78df-11eb-845f-e14e7b5a00dc.png">

**Comment by panzarino** [comment]
@jennifer-shehane Knowing how we look at where to save the commands (without actually testing), I think the issue is that it's not able to save new tests. We currently explicitly look for a `describe` or `context` block - we could probably remove this check and match solely based on line/column but then if the function signature is different it still won't work properly. Perhaps the best way to solve this is to just raise an error that we detected a custom function so we can't directly save but still let the user copy to clipboard.

**Comment by jennifer-shehane** [comment]
@panzarino It did add a new test in the example above, just not exactly in the expected place. I just mean, it didn't completely fail. The copy to clipboard is a solution anyone could use though.

**Comment by panzarino** [comment]
@jennifer-shehane I see now - makes sense that it works that way (I forgot that we actually have an e2e test to ensure that it does work like this). Not sure that there's anything we could/should change to handle this particular situation.

Planning on adding copy to clipboard to the backlog given the popular demand.

**Comment by PetMou** [comment]
I also have the issue that commands are not stored in the source.
![afbeelding](https://user-images.githubusercontent.com/48205462/118973360-e2a77380-b971-11eb-99df-4f90c8a3f8c9.png)
Cypress version 7.3.0, Windows 10 using IntelliJ with a 'fresh' javascript test.
`cypress open --config-file cypress/configs/dev.json --config integrationFolder=cypress/integration_sandbox,watchForFileChanges=false
`
(the fileChange=false was necessary to capture the error message)

**Comment by panzarino** [comment]
@PetMou Could you post the stack trace that you get? In addition it might help to post what your test looks like and your config.

**Comment by panzarino** [comment]
@elio-maggini Are you still experiencing this issue with the latest version of Cypress? If so, is your preprocessor generating proper source maps? I looked into it's source code a bit and (just from looking at the code, not testing) it does not appear to be generating source maps. Studio uses the source map for a file to figure out where to save commands, so without sourcemaps its unable to save (this should be improved in the future so that it does not rely on source maps). Is there a reason that you're using that specific preprocessor? Typescript is supported in the default predecessor which does generate proper sourcemaps.

**Comment by panzarino** [comment]
We've just released copy to clipboard in v7.7.0 - check it out and let us know what you think!

We're still working on a more dedicated fix to prevent this message from occurring in the first place but copy to clipboard should help out in the meantime.

**Comment by elio-maggini** [comment]
Thanks! This combined with strategic use of data-cy tags really helps a lot.
